### ISSUE TITLE
微调后，音频的语言检测结果不准了。

### ISSUE THREAD

**Issue by charleybin** [yeyupiaoling/Whisper-Finetune#88]
问题描述：
返回结果中， dataset/test.wav的识别结果为英文内容。同时返回的检测结果显示为 'language': 'english'

具体信息：
python lang-detect.py --audio_path=dataset/test.wav --model_path=models/whisper-large-v2-finetune/
Loading checkpoint shards: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00,  6.70it/s]
加载音频文件：dataset/test.wav
/home/work/.conda/envs/whisper-finetune/lib/python3.10/site-packages/transformers/models/whisper/generation_whisper.py:496: FutureWarning: The input name `inputs` is deprecated. Please make sure to use `input_features` instead.
  warnings.warn(
Due to a bug fix in https://github.com/huggingface/transformers/pull/28687 transcription using a multilingual Whisper will default to language detection followed by transcription instead of translation to English.This might be a breaking change for your use case. If you want to instead always translate your audio to English, make sure to pass `language='en'`.
Passing a tuple of `past_key_values` is deprecated and will be removed in Transformers v4.43.0. You should pass an instance of `EncoderDecoderCache` instead, e.g. `past_key_values=EncoderDecoderCache.from_legacy_cache(past_key_values)`.
The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
{'text': ' In recent years, I have not only crushed my daughter with books, but also persuaded my relatives not to give her crushed money, but to give her crushed books instead.', 'chunks': [{'language': 'english', 'text': ' In recent years, I have not only crushed my daughter with books, but also persuaded my relatives not to give her crushed money, but to give her crushed books instead.'}]}
###########################################################################
[' In recent years, I have not only crushed my daughter with books, but also persuaded my relatives not to give her crushed money, but to give her crushed books instead.']
['<|en|>']
tensor([0.3586], device='cuda:0')

期望结果：
中文语音wav的检测结果为中文。
英文语音wav的检测结果为英文。
其它语言文件的检测结果也能正确返回。

**Comment by charleybin** [comment]
code of lang-detect.py

```Python
import time
from typing import Optional, Collection, List, Dict
import torchaudio
import argparse
import functools
import torch
from transformers import pipeline, WhisperForConditionalGeneration, WhisperTokenizer, WhisperProcessor

parser = argparse.ArgumentParser(description=__doc__)
add_arg = functools.partial(add_arguments, argparser=parser)
add_arg("audio_path",  type=str,  default="dataset/test.wav", help="预测的音频路径")
add_arg("model_path",  type=str,  default="vrv-models/whisper-large-v2-finetune-finetune/", help="合并模型的路径，或者是huggingface上模型的名称")
args = parser.parse_args()

device = "cuda:0" if torch.cuda.is_available() else "cpu"
torch_dtype = torch.float16 if torch.cuda.is_available() else torch.float32

model = WhisperForConditionalGeneration.from_pretrained(args.model_path)
model.to(device)
tokenizer = WhisperTokenizer.from_pretrained(args.model_path, device=device)
processor = WhisperProcessor.from_pretrained(args.model_path, device=device)

pipe = pipeline(
    "automatic-speech-recognition",
    model=model,
    tokenizer=processor.tokenizer,
    feature_extractor=processor.feature_extractor,
    torch_dtype=torch_dtype,
    device=device,
    return_language=True
)
result = pipe(args.audio_path)

print(result)
```

**Comment by ILG2021** [comment]
微调的时候要指定语言

**Comment by charleybin** [comment]
finetune.py 默认的language参数是Chinese，这个没有动哦。

**Comment by yeyupiaoling** [comment]
@charleybin 你微调的是中文数据集吗？

**Comment by jiaohuix** [comment]
我也是遇到了语言不准的情况，我微调多语言的数据集，微调时候language设置为None，数据集里面指定了语言：
```
{"audio": {"path": "/root/zjh/Projects/ml/challenge/asr/data/ify24_5lang/v1/spain/TEDX_F_023_SPA_0019.wav"}, "sentence": "usando un unfirme aburrido escribiendo en el sótano de un hotel literal para alguien que no tenía ni idea de que yo estaba ahí", "language": "Spanish", "duration": 9.8551875, "sentences": [{"start": 0, "end": 9.8551875, "text": "usando un unfirme aburrido escribiendo en el sótano de un hotel literal para alguien que no tenía ni idea de que yo estaba ahí"}]}
{"audio": {"path": "/root/zjh/Projects/ml/challenge/asr/data/ify24_5lang/v1/rus/f99e4416594c.wav"}, "sentence": "им надо учиться в школе", "language": "Russian", "duration": 2.8790625, "sentences": [{"start": 0, "end": 2.8790625, "text": "им надо учиться в школе"}]}
{"audio": {"path": "/root/zjh/Projects/ml/challenge/asr/data/ify24_5lang/v1/spain/TEDX_M_022_SPA_0004.wav"}, "sentence": "en en en esta parte tan particular que son las las chinampas y los corredores de agua y me di cuenta de toda la problematica que hay en la e en la agricultura", "language": "Spanish", "duration": 9.387125, "sentences": [{"start": 0, "end": 9.387125, "text": "en en en esta parte tan particular que son las las chinampas y los corredores de agua y me di cuenta de toda la problematica que hay en la e en la agricultura"}]}
{"audio": {"path": "/root/zjh/Projects/ml/challenge/asr/data/ify24_5lang/v1/ja/sfw-00023-of-00039/fem_hiy_00231_c.wav"}, "sentence": "妃愛は猫の鳴き真似をしました！", "language": "Japanese", "duration": 2.26, "sentences": [{"start": 0, "end": 2.26, "text": "妃愛は猫の鳴き真似をしました！"}]}
{"audio": {"path": "/root/zjh/Projects/ml/challenge/asr/data/ify24_5lang/v1/arabic/common_voice_ar_24168156.wav"}, "sentence": "مشى إلى البيت.", "language": "Arabic", "duration": 2.7075, "sentences": [{"start": 0, "end": 2.7075, "text": "مشى إلى البيت."}]}

```
训练结束后用ct2翻译，语言都成了kn：
![5bd33de4e1783ecc123d28db0ab8119](https://github.com/user-attachments/assets/3f770727-1918-4d55-b7ac-02b267699ac7)

而且预测的结果不同语言混杂，且出现了转录任务与翻译任务混淆的情况，例如阿拉伯asr结果为英文，用原始阿拉伯的转录结果翻译后和asr的结果相近：

![cf22a97c4babf3512210912ab336da9](https://github.com/user-attachments/assets/7d731eaa-1955-4762-93b6-2568503a4c7b)
![f23eaff65f48bef8a7a81a7316505c1](https://github.com/user-attachments/assets/00ab88f9-c92c-447f-bfd7-f4e59e244579)

这个应该是什么问题呀，跪求飘零大佬求解惑~

**Comment by ericyue** [comment]
我也遇到这个问题 finetune后 语言识别不准  手动指定lang后就可以，不指定就很差

**Comment by jiaohuix** [comment]
@ericyue 我大概定位到了问题，但是没解决，应该是ctranslate2转换或者预测时候出现了问题。使用evaluation脚本里面hf的WhisperForConditionalGeneration能够得到正常的结果，训练的时候开始几个token也是正常的: [bos_token, lang_token, task_token, timestamp_token]。
哪位大佬有解决的方法嘛

**Comment by ericyue** [comment]
> @ericyue 我大概定位到了问题，但是没解决，应该是ctranslate2转换或者预测时候出现了问题。使用evaluation脚本里面hf的WhisperForConditionalGeneration能够得到正常的结果，训练的时候开始几个token也是正常的: [bos_token, lang_token, task_token, timestamp_token]。 哪位大佬有解决的方法嘛

应该不是ct2的问题。 你试一下直接finetune merge后的模型， 不指定language识别一下？ 感觉是finetune的问题

**Comment by XM233333333** [comment]
我的也是微调了中文数据集，微调后的模型用代码infer.py推理中文音频，输出了英语。没有指定language，但是不应该啊，怎么解决？

**Comment by yeyupiaoling** [comment]
@XM233333333 

>我的也是微调了中文数据集，微调后的模型用代码 infer.py 推理中文音频，输出了英语。没有指定语言，但是不应该啊，怎么解决？

你用的是tiny模型吗？有时候微调，推理的是翻译的内容。

**Comment by jiaohuix** [comment]
感觉微调过后模型对语言识别变差了，infer时候需要手动指定语言好点

**Comment by ericyue** [comment]
> 感觉微调过后模型对语言id识别变差了，infer时候需要手动指定语言好点

如果我的场景预先不知道语言类型要怎么办呢？比如可能中文也可能英文，或者混合。 有什么好办法
